core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2380,10 +2381,14 @@ class SparkContext(config: SparkConf) extends Logging {
 
   /** Reports heartbeat metrics for the driver. */
   private def reportHeartBeat(): Unit = {
-    val driverUpdates = _heartbeater.getCurrentMetrics()
+    val currentMetrics = ExecutorMetrics.getCurrentMetrics(env.memoryManager)
+    val driverUpdates = new HashMap[(Int, Int), ExecutorMetrics]
+    // In the driver, we do not track per-stage metrics, so use a dummy stage
+    // for the key
+    driverUpdates.put((-1, -1), new ExecutorMetrics(currentMetrics))
```

Comment:
The driver case is a bit different, because we do not know what stage(s) is/are running when we poll the metrics, so we just use a dummy key to keep the format of the update the same, a Map[(Int, Int), ExecutorMetrics]. In the EventLoggingListener, in onExecutorMetricsUpdate, if an update came from the driver, we record those peaks for all active stages. Thus, we do get per stage metrics from the driver in the event logs.

Reply:
Yes, in onExecutorMetricsUpdate the stage information is added, so not needed here.